packages/lodestar/src/sync/backfill/backfill.ts

```diff
@@ -0,0 +1,220 @@
+    }
+
+    // Attempt to find the anchor block in the DB
+    try {
+      this.anchorBlock = await this.db.blockArchive.getByRoot(root);
+    } catch (e) {
+      this.logger.error("Error fetching blockArchive", {root: toHexString(root)}, e);
+    }
+
+    // Start processor loop
+    this.processor.trigger();
+
+    for await (const _ of this.processor) {
+      const peer = shuffleOne(this.peers.values());
+      if (!peer) {
+        this.logger.debug("BackfillSync no peers yet");
+        continue;
+      }
+
+      try {
+        // First get anchor block
+        if (!this.anchorBlock) {
+          const anchorBlockRoot = this.chain.getFinalizedCheckpoint().root;
+          await this.syncBlockByRoot(peer, anchorBlockRoot);
+          if (this.anchorBlock) {
+            this.logger.debug("BackfillSync fetched anchorBlock", {root: toHexString(anchorBlockRoot)});
+          } else {
+            // If syncBlockByRoot could not find anchorBlock, loop and try again with another peer
+            continue;
+          }
+        }
+
+        const oldestSlotRequired = this.getOldestRequiredSlot();
+        const anchorBlockSlot = this.anchorBlock.message.slot;
+        if (anchorBlockSlot <= oldestSlotRequired) {
+          // Sync completed, return the oldestSlotSynced
+          return anchorBlockSlot;
+        }
+
+        const toSlot = anchorBlockSlot;
+        const fromSlot = Math.max(toSlot - this.opts.batchSize, oldestSlotRequired);
+        this.logger.debug("BackfillSync syncing range", {fromSlot, toSlot});
+        await this.syncRange(peer, fromSlot, toSlot, this.anchorBlock.message.parentRoot);
+      } catch (e) {
+        this.logger.debug("BackfillSync sync error", e);
+
+        if (e instanceof BackfillSyncError) {
+          switch (e.type.code) {
+            case BackfillSyncErrorCode.INVALID_SIGNATURE:
+            case BackfillSyncErrorCode.NOT_LINEAR:
+            case BackfillSyncErrorCode.NOT_ANCHORED:
+              this.network.reportPeer(peer, PeerAction.LowToleranceError, "BadSyncBlocks");
+          }
+        }
+      } finally {
+        this.processor.trigger();
+      }
+    }
+
+    throw new ErrorAborted("BackfillSync");
+  }
+
+  private getOldestRequiredSlot(): number {
+    const oldestRequiredEpoch = Math.max(
+      GENESIS_EPOCH,
+      this.chain.clock.currentEpoch - getMinEpochForBlockRequests(this.config)
+    );
+    return computeStartSlotAtEpoch(oldestRequiredEpoch);
+  }
+
+  private addPeer = (peerId: PeerId, peerStatus: phase0.Status): void => {
+    const requiredSlot =
+      this.anchorBlock?.message.slot ?? computeStartSlotAtEpoch(this.chain.getFinalizedCheckpoint().epoch);
+    this.logger.debug("BackfillSync add peer", {peerhead: peerStatus.headSlot, requiredSlot});
+    if (peerStatus.headSlot >= requiredSlot) {
+      this.peers.add(peerId);
+      this.processor.trigger();
+    }
+  };
+
+  private removePeer(peerId: PeerId): void {
+    this.peers.delete(peerId);
```

Comment:
We cannot report peers here, otherwise we would downscore peers just for disconnecting.